### ISSUE TITLE
IoU localization loss function doesn't work with object detection - Wrong tensor format inside function

### ISSUE THREAD

**Issue by lernerbruno** [tensorflow/models#7545]
### System information
- **What is the top-level directory of the model you are using**: object_detection
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Mac  OS 10.13.3
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: tf 1.9.0
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**: 
- **Exact command to reproduce**:

### Describe the problem

One option of localization loss provided is the **IOU loss**.    

When I first tried to use it, I got a problem with the shapes.. it reshapes the loss output to a 1D array, and this is not what the rest of the code expects, so I've reshaped to [batch_size, num_anchors, -1] and it worked, but the values were not converging.. so I went to further 
investigation.     
 
The function that calculates the IoU (`matched_iou()`) expects the boxes in a **decoded format** (`[Xmin, Ymin, Xmax, Ymax]`), but if you print the prediction and target tensors, they come in the **encoded format** (according to the default boxes). In order to decode it, you need to use the default boxes, which at this point in the code are not there anymore.

### Source code / logs
Code changed by me in the IoU Localization loss class in `object_detection/core/losses.py`
```
    def _compute_loss(self, prediction_tensor, target_tensor, weights):
        """Compute loss function.

        Args:
          prediction_tensor: A float tensor of shape [batch_size, num_anchors, 4]
            representing the decoded predicted boxes
          target_tensor: A float tensor of shape [batch_size, num_anchors, 4]
            representing the decoded target boxes
          weights: a float tensor of shape [batch_size, num_anchors]

        Returns:
          loss: a float tensor of shape [batch_size, num_anchors] tensor
            representing the value of the loss function.
        """

        prediction_tensor = tf.Print(prediction_tensor, [prediction_tensor], 'Prediction tensor in IOU loss: ', summarize=8)
        target_tensor = tf.Print(target_tensor, [target_tensor], 'Target tensor in IOU loss: ', summarize=8)

        shape = prediction_tensor.get_shape().as_list()
        batch_size, num_anchors, _ = shape

        predicted_boxes = box_list.BoxList(tf.reshape(prediction_tensor, [-1, 4]))
        target_boxes = box_list.BoxList(tf.reshape(target_tensor, [-1, 4]))
        per_anchor_iou_loss = 1.0 - box_list_ops.matched_iou(predicted_boxes,
                                                             target_boxes)

        return tf.reshape(weights, [batch_size, num_anchors, -1]) * tf.reshape(per_anchor_iou_loss,\
                                                                               [batch_size, num_anchors, -1])

```
Output:
```
Prediction tensor in IOU loss: [[[0.91901803 1.04905343 0.639230669 -0.0775884241][-0.474719614 0.0712455213 -0.291270375 1.18328691]]...]
Target tensor in IOU loss: [[[0 0 0 0][0 0 0 0]]...]
Prediction tensor in IOU loss: [[[0.611483 1.27656579 0.435161799 0.46589154][-0.00451424718 0.386824429 -0.291154832 0.644339323]]...]
Target tensor in IOU loss: [[[0 0 0 0][0 0 0 0]]...]
Prediction tensor in IOU loss: [[[-0.0916902 0.887826204 -0.670153 0.0180809647][-1.30561447 -0.733022928 -0.626737714 0.740641832]]...]
Target tensor in IOU loss: [[[0 0 0 0][0 0 0 0]]...]
```

**Comment by shsjxzh** [comment]
@lernerbruno Thank you for identifying the problem! Here is a work around if you need:
I will take SSD architecture as an example. 

First, Open file "object_detection/meta_architectures/ssd_meta_arch.py", and add this line at the top of the file:
```python
from object_detection.core import losses
```

Then, add and replace a few lines of code at "loss" function:
```python
if self.groundtruth_has_field(fields.InputDataFields.is_annotated):
  losses_mask = tf.stack(self.groundtruth_lists(
      fields.InputDataFields.is_annotated))

# TODO: start from here to fix the bug!!

if isinstance(self._localization_loss, losses.WeightedIOULocalizationLoss) :
  anchors = self.anchors.get()
  batch_reg_targets, _ = self._batch_decode(batch_reg_targets, anchors)
  predict_decoded, _ = self._batch_decode(prediction_dict['box_encodings'], anchors)
else:
  predict_decoded = prediction_dict['box_encodings']

location_losses = self._localization_loss(
  predict_decoded,
  batch_reg_targets,
  ignore_nan_targets=True,
  weights=batch_reg_weights,
  losses_mask=losses_mask)
```

With these fixes, you can train you networks with iou loss. :smile:

**Comment by icoicqico** [comment]
Hello, thanks for the fixes. I can start training with the fixes you provided, but the loss div to Nan, and training stop in around 1k steps.

**Comment by shsjxzh** [comment]
IOU loss should have a different learning rate setting from the generally used smooth L1 loss. You can try to turn down the learning rate a little.
